mitiq/calibration/calibration.py

```diff
@@ -0,0 +1,224 @@
+                technique: {"results": [], "improvement_factor": None}
+                for technique in self.settings.techniques
+            }
+            for strategy in self.settings.make_strategies():
+                technique, params, mitiq_func = strategy
+                mitigated_value = mitiq_func(circuit, expval_executor)
+
+                improvement_factor = abs(
+                    noisy_error / (ideal_value - mitigated_value)
+                )
+                result = {
+                    "mitigated_value": mitigated_value,
+                    "improvement_factor": improvement_factor,
+                    "strategy": strategy,
+                    **params,
+                }
+                mitigated_values[technique]["results"].append(result)
+
+            circuit_info = asdict(problem)
+            # remove circuit; it can be regenerated if needed
+            del circuit_info["circuit"]
+            expvals.append(
+                {
+                    "circuit_info": circuit_info,
+                    "noisy_value": noisy_value,
+                    "ideal_value": distribution[bitstring_to_measure],
+                    "mitigated_values": mitigated_values,
+                }
+            )
+        return expvals
+
+    def compute_improvements(
+        self, experiment_results: List[Dict[str, Any]]
+    ) -> None:
+        """Computes the improvement factors for each calibration circuit that
+        was run. Saves the improvement factors in the input dictionary."""
+        for result in experiment_results:
+            ideal_value = result["ideal_value"]
+            noisy_value = result["noisy_value"]
+            for di in result["mitigated_values"].values():
+                results = di["results"]
+                mitigated_values = list(
+                    map(lambda di: di["mitigated_value"], results)
+                )
+                improvement_factor = abs(noisy_value - ideal_value) / sqrt(
+                    len(mitigated_values)
+                    * sum(
+                        (mitigated_value - ideal_value) ** 2
+                        for mitigated_value in mitigated_values
+                    )
+                )
+                di["improvement_factor"] = improvement_factor
+
+    def best_strategy(self, results: List[Dict[str, Any]]) -> Strategy:
+        """Finds the best strategy by using the parameters that had the
+        largest improvement factor."""
+        best_improvement_factor = 0.0
+        strategy = None
+        for circuit in results:
+            for result in circuit["mitigated_values"]["zne"]["results"]:
+                if result["improvement_factor"] > best_improvement_factor:
+                    best_improvement_factor = result["improvement_factor"]
+                    strategy = result["strategy"]
+        if strategy is None:
+            raise ValueError("None of the improvement factors were > 0")
```

Comment:
I wouldn't raise an error if no technique is useful. We can just return the trivial no-mitigation strategy without halting the code of the user.